lib/std/simd.zig

```diff
@@ -0,0 +1,395 @@
+        out[i] = @shuffle(Child, interlaced, undefined, indices);
+    }
+
+    return out;
+}
+
+pub fn extract(
+    vec: anytype,
+    comptime first: VectorIndex(@TypeOf(vec)),
+    comptime count: VectorCount(@TypeOf(vec)),
+) Vector(count, std.meta.Child(@TypeOf(vec))) {
+    const Child = std.meta.Child(@TypeOf(vec));
+    const len = vectorLength(@TypeOf(vec));
+
+    std.debug.assert(@intCast(comptime_int, first) + @intCast(comptime_int, count) <= len);
+
+    return @shuffle(Child, vec, undefined, iota(i32, count) + @splat(count, @intCast(i32, first)));
+}
+
+test "vector patterns" {
+    const base = Vector(4, u32){ 10, 20, 30, 40 };
+    const other_base = Vector(4, u32){ 55, 66, 77, 88 };
+
+    const small_bases = [5]Vector(2, u8){
+        Vector(2, u8){ 0, 1 },
+        Vector(2, u8){ 2, 3 },
+        Vector(2, u8){ 4, 5 },
+        Vector(2, u8){ 6, 7 },
+        Vector(2, u8){ 8, 9 },
+    };
+
+    try std.testing.expectEqual([6]u32{ 10, 20, 30, 40, 10, 20 }, repeat(6, base));
+    try std.testing.expectEqual([8]u32{ 10, 20, 30, 40, 55, 66, 77, 88 }, join(base, other_base));
+    try std.testing.expectEqual([2]u32{ 20, 30 }, extract(base, 1, 2));
+
+    if (comptime !builtin.cpu.arch.isMIPS()) {
+        try std.testing.expectEqual([8]u32{ 10, 55, 20, 66, 30, 77, 40, 88 }, interlace(.{ base, other_base }));
+
+        const small_braid = interlace(small_bases);
+        try std.testing.expectEqual([10]u8{ 0, 2, 4, 6, 8, 1, 3, 5, 7, 9 }, small_braid);
+        try std.testing.expectEqual(small_bases, deinterlace(small_bases.len, small_braid));
+    }
+}
+
+/// Elements are shifted to higher indices. Elements that pass the top are discarded.
+pub fn shiftElementsRight(vec: anytype, comptime amount: VectorCount(@TypeOf(vec)), shift_in: std.meta.Child(@TypeOf(vec))) @TypeOf(vec) {
+    return join(@splat(amount, shift_in), extract(vec, 0, vectorLength(@TypeOf(vec)) - amount));
+}
+
+/// Elements are shifted to lower indices. Elements that pass the bottom are discarded.
+pub fn shiftElementsLeft(vec: anytype, comptime amount: VectorCount(@TypeOf(vec)), shift_in: std.meta.Child(@TypeOf(vec))) @TypeOf(vec) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return join(extract(vec, amount, len - amount), @splat(amount, shift_in));
+}
+
+/// Elements are shifted to lower indices. Elements that pass the bottom return to the top.
+pub fn rotateElementsLeft(vec: anytype, comptime amount: VectorCount(@TypeOf(vec))) @TypeOf(vec) {
+    const Child = std.meta.Child(@TypeOf(vec));
+    const len = vectorLength(@TypeOf(vec));
+
+    std.debug.assert(amount <= len);
+
+    const indices = comptime @mod(iota(i32, len) + @splat(len, @as(i32, amount)), @splat(len, @intCast(i32, len)));
+
+    return @shuffle(Child, vec, undefined, indices);
+}
+
+/// Elements are shifted to higher indices. Elements that pass the top return to the bottom.
```

Comment:
Phrasing could be improved. Maybe something like the following:
-Elements are shifted to higher indices.
+Vector elements are shifted rightwards.
-Elements that pass the top return to the bottom.
+Elements that pass the highest index are returned relative to the offset starting at index 0.

Reply:
I say "higher indices" because left versus right is somewhat ambigious, as discussed earlier. Will "shifted rightwards (towards higher indicies)" work?

Reply:
That would also work. Only relevant things is make it clear what is meant with the terminology. see my comment below ` pub const Vector = std.meta.Vector;` as the basic structure.